File: src/infrastructure/validators/rgp.ts
import { IValidationResult, ok, fail } from './result';
import { isUF } from './uf';

const STRICT = /^[A-Z]{2}-\d{7}-\d$/;
// Tolerante: captura "SC 0004633 8", "sc-0004633/8", etc., evitando colar em letras/dígitos vizinhos.
const LOOSE = /(?<![A-Za-z0-9])([A-Za-z]{2})[\s\-./]*(\d{7})[\s\-./]*(\d)(?!\d)/;

/**
 * Valida e normaliza o número do RGP para o formato `UF-0000000-0`.
 * Aceita variações de separador; sinaliza por mensagem quando precisou normalizar.
 */
export function validateRgp(raw: string | null | undefined): IValidationResult {
  const value = (raw ?? '').trim();
  if (!value) return fail('RGP vazio');

  const upper = value.toUpperCase();

  if (STRICT.test(upper)) {
    const uf = upper.slice(0, 2);
    if (!isUF(uf)) return fail(`UF do RGP inválida: ${uf}`);
    return ok(upper);
  }

  const m = LOOSE.exec(value);
  if (!m) return fail('Formato de RGP não reconhecido (esperado UF-0000000-0)');

  const uf = m[1].toUpperCase();
  if (!isUF(uf)) return fail(`UF do RGP inválida: ${uf}`);

  const normalized = `${uf}-${m[2]}-${m[3]}`;
  return { valid: true, normalized, messages: [`RGP normalizado a partir de formato divergente: "${value}"`] };
}

File: app/validators/rgp.py
"""Validação e normalização do número de RGP.

Formato esperado (Registro Geral da Atividade Pesqueira):
    UF + 7 dígitos + dígito final, ex.: SC-0004633-8
Regex de referência: ^[A-Z]{2}-\\d{7}-\\d$
"""
from __future__ import annotations

import re

from .base import ValidationResult
from .uf import UFS

# Aceita variações de entrada: com/sem hífens, com espaços, minúsculas.
# Lookbehind/lookahead garantem que a UF não absorva letras vizinhas
# (ex.: "ABSC-0004633-8" NÃO deve casar como UF "SC").
_LOOSE = re.compile(r"(?<![A-Za-z0-9])([A-Za-z]{2})[\s\-./]*(\d{7})[\s\-./]*(\d)(?!\d)")
_STRICT = re.compile(r"^[A-Z]{2}-\d{7}-\d$")


def format_rgp(uf: str, body: str, dv: str) -> str:
    return f"{uf.upper()}-{body}-{dv}"


def is_valid(raw: str) -> bool:
    return validate(raw).valid


def validate(raw: str) -> ValidationResult:
    text = (raw or "").strip().upper()
    if not text:
        return ValidationResult(False, None, ["RGP vazio"])

    if _STRICT.match(text):
        uf = text[:2]
        if uf not in UFS:
            return ValidationResult(False, None, [f"UF do RGP desconhecida: {uf}"])
        return ValidationResult(True, text, [])

    match = _LOOSE.search(text)
    if not match:
        return ValidationResult(
            False, None, ["RGP fora do padrão esperado (UF-0000000-0)"]
        )

    uf, body, dv = match.group(1), match.group(2), match.group(3)
    messages: list[str] = []
    if uf not in UFS:
        messages.append(f"UF do RGP desconhecida: {uf}")
        return ValidationResult(False, None, messages)

    normalized = format_rgp(uf, body, dv)
    # Foi necessário "limpar" a entrada -> aviso para conferência humana.
    if normalized != text:
        messages.append("RGP normalizado a partir de formato divergente")
    return ValidationResult(True, normalized, messages)
